contracts/Pool.sol

```diff
@@ -0,0 +1,381 @@
+
+    _updateRates();
+
+    emit Withdrawal(msg.sender, _amount, block.timestamp);
+  }
+
+
+  /**
+   * Borrows the specified amount
+   * It updates user borrowed balance, total borrowed amount and rates
+   * @dev _amount the amount to be borrowed
+  **/
+  function borrow(uint256 _amount) payable external canBorrow nonReentrant {
+    require(address(this).balance >= _amount, "There is not enough funds in the pool to fund the loan.");
+    require(totalSupply() - totalBorrowed() >= _amount, "There is no enough deposit in the pool to fund the loan.");
+
+    _accumulateBorrowingInterest(msg.sender);
+
+    borrowed[msg.sender] += _amount;
+    borrowed[address(this)]+= _amount;
+
+    payable(msg.sender).transfer(_amount);
+
+    _updateRates();
+
+    emit Borrowing(msg.sender, _amount, block.timestamp);
+  }
+
+  /**
+   * Repays the message value
+   * It updates user borrowed balance, total borrowed amount and rates
+  **/
+  function repay() payable external {
+    _accumulateBorrowingInterest(msg.sender);
+
+    require(borrowed[msg.sender] >= msg.value, "You are trying to repay more that was borrowed by user.");
+
+    borrowed[msg.sender] -= msg.value;
+    borrowed[address(this)] -= msg.value;
+
+    _updateRates();
+
+    emit Repayment(msg.sender, msg.value, block.timestamp);
+  }
+
+
+  /* =========
+
+
+  /**
+   * Returns the current borrowed amount for the given user
+   * The value includes the interest rates owned at the current moment
+   * @dev _user the address of queried borrower
+  **/
+  function getBorrowed(address _user) public view returns (uint256) {
+    return borrowIndex.getIndexedValue(borrowed[_user], _user);
+  }
+
+  function totalSupply() public view override returns (uint256) {
+    return balanceOf(address(this));
+  }
+
+  function totalBorrowed() public view returns (uint256) {
+    return getBorrowed(address(this));
+  }
+
+
+  /**
+   * Returns the current deposited amount for the given user
+   * The value includes the interest rates earned at the current moment
+   * @dev _user the address of queried depositor
+  **/
+  function balanceOf(address user) public view override returns (uint256) {
+    return depositIndex.getIndexedValue(_deposited[user], user);
+  }
+
+
+  /**
+   * Returns the current interest rate for deposits
+  **/
+  function getDepositRate() public view returns (uint256) {
+    return _ratesCalculator.calculateDepositRate(totalBorrowed(), totalSupply());
+  }
+
+
+  /**
+   * Returns the current interest rate for borrowings
+  **/
+  function getBorrowingRate() public view returns (uint256) {
+    return _ratesCalculator.calculateBorrowingRate(totalBorrowed(), totalSupply());
+  }
+
+
+  /**
+   * Recovers the surplus funds resultant from difference between deposit and borrowing rates
+  **/
+  function recoverSurplus(uint256 amount, address account) public onlyOwner nonReentrant {
+    uint256 surplus = address(this).balance + totalBorrowed() - totalSupply();
```

Comment:
this need more testing- like multiple recover of surplus with time and checking if pool still balances

Reply:
done